pkg/services/ngalert/state/historian/annotation.go

```diff
@@ -31,48 +31,60 @@ func NewAnnotationHistorian(annotations annotations.Repository, dashboards dashb
+func (h *AnnotationStateHistorian) RecordStates(ctx context.Context, states []state.ContextualState) {
+	go h.recordStatesSync(ctx, states)
+}
 
-	item := &annotations.Item{
-		AlertId:   rule.ID,
-		OrgId:     rule.OrgID,
-		PrevState: previousData.String(),
-		NewState:  currentData.String(),
-		Text:      annotationText,
-		Epoch:     evaluatedAt.UnixNano() / int64(time.Millisecond),
-	}
+func (h *AnnotationStateHistorian) recordStatesSync(ctx context.Context, states []state.ContextualState) {
+	items := make([]annotations.Item, 0, len(states))
+	for _, state := range states {
+		h.log.Debug("alert state changed creating annotation", "alertRuleUID", state.AlertRuleUID, "newState", state.Formatted(), "oldState", state.PreviousFormatted())
 
-	dashUid, ok := rule.Annotations[ngmodels.DashboardUIDAnnotation]
-	if ok {
-		panelUid := rule.Annotations[ngmodels.PanelIDAnnotation]
+		labels := removePrivateLabels(state.Labels)
+		annotationText := fmt.Sprintf("%s {%s} - %s", state.RuleTitle, labels.String(), state.Formatted())
 
-		panelId, err := strconv.ParseInt(panelUid, 10, 64)
-		if err != nil {
-			h.log.Error("error parsing panelUID for alert annotation", "panelUID", panelUid, "alertRuleUID", rule.UID, "err", err.Error())
-			return
+		item := annotations.Item{
+			AlertId:   state.RuleID,
+			OrgId:     state.OrgID,
+			PrevState: state.PreviousFormatted(),
+			NewState:  state.Formatted(),
+			Text:      annotationText,
+			Epoch:     state.LastEvaluationTime.UnixNano() / int64(time.Millisecond),
 		}
 
-		query := &models.GetDashboardQuery{
-			Uid:   dashUid,
-			OrgId: rule.OrgID,
-		}
+		dashUid, ok := state.Annotations[ngmodels.DashboardUIDAnnotation]
+		if ok {
+			panelUid := state.Annotations[ngmodels.PanelIDAnnotation]
 
-		err = h.dashboards.GetDashboard(ctx, query)
-		if err != nil {
-			h.log.Error("error getting dashboard for alert annotation", "dashboardUID", dashUid, "alertRuleUID", rule.UID, "err", err.Error())
-			return
-		}
+			panelId, err := strconv.ParseInt(panelUid, 10, 64)
+			if err != nil {
+				h.log.Error("error parsing panelUID for alert annotation", "panelUID", panelUid, "alertRuleUID", state.AlertRuleUID, "err", err.Error())
+				return
+			}
 
-		item.PanelId = panelId
-		item.DashboardId = query.Result.Id
+			query := &models.GetDashboardQuery{
+				Uid:   dashUid,
+				OrgId: state.OrgID,
+			}
+
+			err = h.dashboards.GetDashboard(ctx, query)
+			if err != nil {
+				h.log.Error("error getting dashboard for alert annotation", "dashboardUID", dashUid, "alertRuleUID", state.AlertRuleUID, "err", err.Error())
+				return
+			}
+
+			item.PanelId = panelId
+			item.DashboardId = query.Result.Id
+		}
+		items = append(items, item)
 	}
 
-	if err := h.annotations.Save(ctx, item); err != nil {
-		h.log.Error("error saving alert annotation", "alertRuleUID", rule.UID, "err", err.Error())
+	if err := h.annotations.SaveMany(ctx, items); err != nil {
```

Comment:
I debated wrapping this with the same feature flag used in https://github.com/grafana/grafana/pull/55350

But, I also remember a Slack message saying that we're getting rid of that flag ASAP, so stayed away from it. WDYT @joeblubaugh? Is there a timeline for removing that flag? It's very relevant here.